test/tests/selftest/pbs_param_dict.py

```diff
@@ -0,0 +1,125 @@
+#
+# This file is part of both the OpenPBS software ("OpenPBS")
+# and the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# OpenPBS is free software. You can redistribute it and/or modify it under
+# the terms of the GNU Affero General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or (at your
+# option) any later version.
+#
+# OpenPBS is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public
+# License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# PBS Pro is commercially licensed software that shares a common core with
+# the OpenPBS software.  For a copy of the commercial license terms and
+# conditions, go to: (http://www.pbspro.com/agreement.html) or contact the
+# Altair Legal Department.
+#
+# Altair's dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of OpenPBS and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair's trademarks, including but not limited to "PBS™",
+# "OpenPBS®", "PBS Professional®", and "PBS Pro™" and Altair's logos is
+# subject to Altair's trademark licensing policies.
+
+
+from tests.selftest import *
+from ptl.utils.plugins.ptl_test_runner import PTLTestRunner
+
+
+class TestParamDict(TestSelf):
+    """
+    Test parsing the pbs_benchpress -p argument
+    """
+
+    def test_gpd(self):
+        """
+        Pass various -p param strings to __get_param_dictionary and check
+        that result is correct.
+        """
+        def try_one(runner, param):
+            """
+            Run get_param_dictionary to convert param string to dict
+            """
+            runner.param = param
+            try:
+                # Note that we call private routine directly to avoid
+                # possible side effects.
+                result = runner._PTLTestRunner__get_param_dictionary()
+            except ValueError:
+                result = None
+            return result
+
+        def compare(old, new, expect_none=False):
+            """
+            Compare new param dict to old
+            """
+            if new is None and not expect_none:
+                return 'Unexpected parse error'
+            if expect_none and new is not None:
+                return 'Should have failed'
+            all_keys = set(old.keys()) | set(new.keys())
+            diffs = []
+            for k in sorted(all_keys):
+                o = old.get(k, None)
+                n = new.get(k, None)
+                if o != n:
+                    diffs.append((k, n))
```

Comment:
Shouldn't you be able to do this with a set intersection?

Reply:
I don't think so, because some of the values are sets and sets are not hashable.

Reply:
I wasn't talking about the values themselves.  More of the keys.  You create a set of the keys of old and new, and then walk down the sorted list checking for differences.  Can you take a set of old.keys() and a set of new.keys(), and do an intersection?

Reply:
You are right--the compare routine was too general purpose. For these tests, we know the old and new dictionaries have the same sets of keys, making it simpler to detect changes.